libass/ass_utils.h

```diff
@@ -182,6 +182,20 @@ static inline int double_to_d22(double x)
     return lrint(x * 0x400000);
 }
 
+static inline int32_t lshiftwrapi(int32_t i, int32_t shift)
+{
+#if UINT32_MAX > INT_MAX
+    // If INT_MAX >= UINT32_MAX, but also INT_MAX < 2^63 - 1
+    // this may still be UB due to auto int promotion
+    return ((uint32_t)i) << (((uint32_t)shift) & 31);
+#else
+    // but this on the other hand is wrong if UINT_MAX < UINT32_MAX.
+    // Extra variable to prevent the compiler from dropping the uint32_t cast.
+    uint32_t tmp = ((unsigned)i) << (((uint32_t)shift) & 31);
+    return tmp;
+#endif
```

Comment:
Could you elaborate on the extra-variable hack? I don’t immediately understand how it can matter.

Keeping in mind that `int32_t` is always two’s complement, I don’t think the cast in `(uint32_t)shift` is necessary, either, is it?

Anyway, how about simplifying all of this to just:
```c
    return 0u + (uint32_t) i << (shift & 31);
```

Reply:
> Could you elaborate on the extra-variable hack? I don’t immediately understand how it can matter.

I'm pretty sure that as far as the spec is concerned, it should be identical to using no extra variable but just a cast to `uintN_t)` *(and before the last push, it did use the latter)*. In practice, testing with `uint8_t`, the compiler apparently to drop both the `unsigned` and `uint8_t` cast and then the sanitizer complains about UB because `left shift of 128 by 25 places cannot be represented in type 'int'` unless the temporary variable is used. Arguably this seems to be more of a compiler/sanitizer bug.

Using your proposed `0u + (uintN_t)i)` instead also makes the compiler and sanitizer happy _and_ allows to drop the preprocessor check, so let’s go with this instead :)